### PR TITLE
6 transaction manager class

### PR DESCRIPTION
**Pull request by junajan**
Add transaction manager class, add basic functions for products

### CODE REVIEW

File: lib/services/product-manager.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+      const productBatches = Promise.map(
+        _.chunk(skus, this.loadBatchCount), // load max N skus per one request
+        skuBatch => this.getProductsBySku(skuBatch),
+        { concurrency: this.loadConcurrency } // load products with concurrency
+      )
+
+      return this.filterOutDuplicateProducts(_.flatten(productBatches))
+    }
+
+    const skuPredicate = `${skus.join('","')}`
```
Comment (lojzatran) [inline lib/services/product-manager.js]: Can be `const skuPredicate = skus.join('","')`

File: lib/services/product-manager.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    return Promise.resolve()
+  }
+
+  deleteProduct (product) {
+    return this.client.products
+      .byId(product.id)
+      .delete(product.version)
+  }
+
+  removeVariantsFromProduct (product, variantSkus) {
```
Comment (lojzatran) [inline lib/services/product-manager.js]: How do you think this should be implemented? If we're going to execute the action here, then how are we going to save it in the custom object backup? 😕  The same question applies for method `anonymizeProduct`.

Comment (junajan) [reply lib/services/product-manager.js]: you are right .. it will be all in runUpdateActions - see the proposed concept of transactions https://gist.github.com/junajan/b4486d98a3c2b2ddc54a0e8f801d8b3c

But we should also cover promoting of new masterVariant and add a check which will decide whether to run action or not - for example, we remove variant v1 which was a masterVariant so we should promote a new one as a masterVariant.

But before we run this some change can happen and the product won't have any other variant or it won't be a masterVariant anymore so we should not call that promoting of new masterVariant. Or we will get a `concurrent modification` error so we recalculate all actions based on the new state and run it again?

Comment (lojzatran) [reply lib/services/product-manager.js]: I'm thinking should we save the actions or maybe we can just save the product draft? Because in case of an error and you need to refetch from the custom object, you may need to recalculate everything anyway?
Or if we save all the actions, then the benefit is only that we don't need to recalculate and can execute the actions as they are with current version.

Comment (lojzatran) [reply lib/services/product-manager.js]: Ah no, it's not enough in case we delete some variants, we might miss some information.

Comment (junajan) [reply lib/services/product-manager.js]: I think that we will have to recalculate everytime we start transaction because the products on API could have been changed so maybe it is a good idea to save just a productDraft and variants which are being removed/moved

Comment (butenkor) [reply lib/services/product-manager.js]: I wonder why product would change (i am ok with inventories change -> version change)? I think we need to state clearly as part of documentation that any process which changes product data needs to make sure that there are no pending operations in TRANSACTION_CONTAINER and only continue if it is empty. I do not like this dependency but otherwise we might run into weird scenarios:

-> actions [{update P1 and V1 attribute to value 10}]
-> another process would change P1 and V1 attribute to value 20
-> we rerun later actions from 1 step and override to 10 again -> wrong state!

Comment (butenkor) [reply lib/services/product-manager.js]: Also we should consider that we raise probability of hitting the limit of 16MB per resource if we save to much data.

Comment (lojzatran) [reply lib/services/product-manager.js]: OK, we will state it clearly in the docs what you wrote. However, from the code perspective it seems easier and less code to write if we recalculate the actions from productDraft saved in custom object than saving the actions itself. Because if we save actions, then we need to add different codes to check if this action has happened etc. 
And as to the 16 mb limit, both approaches could reach the limit with the same probability. We discussed with Jan and came to conclusion that we can add compression to the object. https://github.com/commercetools/commercetools-node-variant-reassignment/issues/10

File: lib/services/product-manager.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+      .fetch()
+      .then(res => res.body.results)
+  }
+
+  getProductById (id) {
+    return this.client.productProjections
+      .staged(true)
+      .where(`id="${id}"`)
+      .fetch()
+      .then(res => res.body.results[0])
```
[the remark this answers is not in the document]
Comment (junajan) [reply lib/services/product-manager.js]: Updated bf4c30f

File: lib/services/product-manager.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      .byId(product.id)
+      .update(request)
+  }
+
+  filterOutDuplicateProducts (products) {
+    console.log(products)
+    throw new Error('NOT IMPLEMENTED')
+  }
+
+  async getProductsBySku (skus) {
```
[the remark this answers is not in the document]
Comment (junajan) [reply lib/services/product-manager.js]: I added `_.uniq` just to be sure bf4c30f

File: lib/services/ctp.js [hunk 5]
```diff
@@ -32,15 +32,15 @@ export function loadCredentials (logger, projectKey) {
 }
 
 /**
- * Will create SphereClient
+ * Will create ctp client
  * @param credentials JSON object with configuration
  */
 export function createClient (credentials) {
   return new SphereClient({ config: credentials })
```
[the remark this answers is not in the document]
Comment (junajan) [reply lib/services/ctp.js]: This `ctp.js` will be used only from tests so I moved it to test utils folder bf4c30f

File: lib/services/product-manager.js (hunk 1 above)
Comment (junajan) [reply lib/services/product-manager.js]: Updated bf4c30f

File: lib/services/product-manager.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+      return this.filterOutDuplicateProducts(_.flatten(productBatches))
+    }
+
+    const skuPredicate = `${skus.join('","')}`
+    const predicate = `masterVariant(sku IN("${skuPredicate}"))`
+      + ` OR variants(SKU IN("${skuPredicate}")`
+
+    return this.client.productProjections
+      .where(predicate)
+      .fetch()
```
[the remark this answers is not in the document]
Comment (junajan) [reply lib/services/product-manager.js]: `this.loadBatchCount` is a constant which will be set by us so we can ensure that the limit is below 501 items per batch.

File: lib/services/product-manager.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (junajan) [reply lib/services/product-manager.js]: Updated bf4c30f

File: lib/services/product-manager.js [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+    if (product)
+      return this.client.products
+        .byId(product.id)
+        .delete(product.version)
+
+    return Promise.resolve()
+  }
+
+  deleteProduct (product) {
```
[the remark this answers is not in the document]
Comment (junajan) [reply lib/services/product-manager.js]: Updated here ec4bb01

File: lib/services/product-manager.js (hunk 2 above)
Comment (junajan) [reply lib/services/product-manager.js]: We have two options - calculate actions which will be executed in specified order but then if/when the process crashes we will have to rerun those actions again but some of them (eg. removing variant from a product) can be already invalid so we will have to check if the actions is still a valid one. So we have to implement action scheduler for calculating actions and then some action validator.

Or we can take productDraft and removed variants from an unfinished transaction and use scheduler again to calculate valid update actions.

Comment (butenkor) [reply lib/services/product-manager.js]: I guess it is ok if save product draft.

File: lib/services/product-manager.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (butenkor) [reply lib/services/product-manager.js]: How will we ensure that? There is no comment or TODO which explains the limitation. New dev might just change it and run into issue on production only.

File: spec/integration/product-manager.spec.js [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+  beforeEach(() =>
+    productService = new ProductManager(utils.logger, ctpClient)
+  )
+
+  afterEach(() =>
+    utils.deleteResource(ctpClient.products)
+  )
+
+  describe('Basic functions', () => {
```
Comment (butenkor) [inline spec/integration/product-manager.spec.js]: I would not define all those tests in this file as integration but rather unit tests so save API calls and time. Those methods will never be used on their own but as whole in particular business use cases and those business use cases should be tested as integration, wdyt?

File: spec/integration/transaction-manager.spec.js [hunk 9]
```diff
@@ -0,0 +1,98 @@
+import _ from 'lodash'
+import { expect } from 'chai'
+
+import * as utils from '../utils/helper'
+import * as constants from '../../lib/constants'
+import TransactionManager from '../../lib/services/transaction-manager'
+
+describe('TransactionManager', () => {
```
Comment (butenkor) [inline spec/integration/transaction-manager.spec.js]: Also here for all below tests i would go with unit instead of integration tests.

File: lib/services/product-manager.js [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+  getProductById (id) {
+    return this.client.productProjections
+      .staged(true)
+      .byId(id)
+      .fetch()
+      .then(res => res.body)
+      .catch(err => (
+        err && err.body && err.body.statusCode
+          ? Promise.resolve(undefined)
+          : Promise.reject(err)
```
[the remark this answers is not in the document]
Comment (junajan) [reply lib/services/product-manager.js]: I don't know why but there is missing something .. I updated it here 39aed15

File: spec/integration/product-manager.spec.js (hunk 8 above)
Comment (junajan) [reply spec/integration/product-manager.spec.js]: I think that the business use cases should be definitely covered by integration tests and a vast majority of service methods should be covered by unit tests (at least where it makes sense).

But in more complicated methods (eg. [removing variants from a product](https://github.com/commercetools/commercetools-node-variant-reassignment/pull/15/files#diff-15ce2e3c2d77d574a8c8ae56d19b9d92R105)) can be plenty of edge cases which can break in the future (for example an order of API commands matters - removing current variant after staged works but not in the reversed order and etc.).

With one or two business tests you won't cover all these edge cases. Also, it does not make sense to write multiple business integration tests and repeatedly test other parts of the system (eg. loading products by SKU) to test the different behavior of `removeVariantsFromProduct`method.

Another reason for integration tests is that sometimes we can reuse services somewhere else so it is nice to have one fully tested and independent component which can be just copied with all tests.

Therefore I would make also some integration tests for services.

But I agree that there are some pointless tests which can be removed so I created this issue which should be taken as one of the last - https://github.com/commercetools/commercetools-node-variant-reassignment/issues/16

File: spec/integration/transaction-manager.spec.js (hunk 9 above)
Comment (junajan) [reply spec/integration/transaction-manager.spec.js]: As in my previous comment https://github.com/commercetools/commercetools-node-variant-reassignment/pull/9#discussion_r154577140 I would like to leave the tests there and we can delete them once we have some other tests which cover these use cases.

File: lib/services/product-manager.js (hunk 2 above)
Comment (butenkor) [reply lib/services/product-manager.js]: @lojzatran @junajan Please provide issue or clear documentation as stated here: https://github.com/commercetools/commercetools-node-variant-reassignment/pull/9#discussion_r152911526

Comment (junajan) [reply lib/services/product-manager.js]: @butenkor I added a documentation issue here https://github.com/commercetools/commercetools-node-variant-reassignment/issues/17
